llvm/include/llvm/Support/PropertySetIO.h

```diff
@@ -188,6 +188,7 @@ class PropertySetRegistry {
       "SYCL/specialization constants default values";
   static constexpr char SYCL_DEVICELIB_REQ_MASK[] = "SYCL/devicelib req mask";
   static constexpr char SYCL_KERNEL_PARAM_OPT_INFO[] = "SYCL/kernel param opt";
+  static constexpr char SYCL_PROGRAM_METADATA[] = "SYCL/program metadata";
```

Comment:
I have a concern here that we have a number of property sets and I think that we could reduce that number by generalizing them. Another concern is that there is not enough design documents for those property sets, which would describe the original intent of those sets, how are they supposed to be used and extended, etc.

In particular, we have `SYCL/misc properties`, `SYCL/assert used` was proposed [here (#3461 comment)]( https://github.com/intel/llvm/pull/3461/files#r633425752) and `SYCL/requirements` is being discussed in #3781.

I'm not asking to change anything right now or even add that documentation I was talking about, because it seems to me that #3781 proposes an interface which is generic enough to be used here as well. Could you please take a look at that PR to see if you have any concerns/objections or any other feedback about proposed `sycl-post-link` changes?

To me it seems that we should be able to replace property set suggested here with one proposed in #3781 just fine, but I want to double-check that with you; and to be on the same page about the fact that we might need to re-work this part later as codebase evolves.

Note: I'm not sure that it makes a lot of sense to wait for that PR or change this PR to make it closer to `SYCL/requirements` format, because the review there is in its early stages and I think it is just better to wait until it finalizes and later refactor everything at once (I mean, property sets).

Reply:
Thank you for pointing me to this. Yes, I completely agree that this could as well be moved to `SYCL/requirements` as soon as it is in there. That solution would also remove the need for the hacky `@` delimiter in the program metadata. It shouldn't be too difficult to transition it once `SYCL/requirements` is there.

I was hoping to one day use the program metadata for listing kernel names as well, as the CUDA driver API isn't great for that ([hence this comment](https://github.com/intel/llvm/pull/3735/files#diff-7525901710934f7bdb2ad36238c4b67163f112d3bd233db7af0b0078b5b01e80R569)) but it could potentially be in `SYCL/misc`, though that doesn't solve the problem of how it would reach the backend.

Reply:
An option would be to keep the approach for passing metadata to the PI backends, but make a more sophisticated data structure for it that can keep both requirements and whatever additional metadata is passed in properties. Whether or not the additional information should be kept in `SYCL/misc` or `SYCL/program metadata` I am not sure, but I can see benefits in both.

Reply:
> I was hoping to one day use the program metadata for listing kernel names as well, as the CUDA driver API isn't great for that (hence this comment)

We should already have this information, if I'm not mistaken, it is propagated through `Entries` section of device image descriptor:
https://github.com/intel/llvm/blob/f6569d1dc8430f6490c9b9be1ffdcfa6d12c7e5b/sycl/include/CL/sycl/detail/pi.h#L727-L729

Reply:
Awesome! I will keep that in mind, thanks. That would make it a lot easier.